File: examples/cpp/Example02HelloWorld.cpp
#include "maf/maf.hpp"

#include <mpi.h>

class HelloWorldAction : public maf::Action, public maf::Transmittable<HelloWorldAction> {

public:

    std::string my_string;
    int my_int;
    float my_float;
    double my_double;

    HelloWorldAction() : maf::Action() {

    };

    HelloWorldAction(std::string string_, int int_, float float_, double double_) : maf::Action() {
        this->my_string = string_;
        this->my_int = int_;
        this->my_float = float_;
        this->my_double = double_;
    };

    void run() override {
        maf::log("Hello World! (from c++)",
                 "     my_string: `", this->my_string,
                 "`     my_int: `", this->my_int,
                 "`     my_float: `", this->my_float,
                 "`     my_double: `", this->my_double, "`"
                );
    };

    void serialize(std::shared_ptr<maf::Archive> archive) override {
        (*archive) & this->my_string;
        (*archive) & this->my_int;
        (*archive) & this->my_float;
        (*archive) & this->my_double;
    };

};


static std::vector<std::shared_ptr<maf::Action>> GetActions(int action_count) {
    std::vector<std::shared_ptr<maf::Action>> actions;

    for (int ii = 0; ii < action_count; ii++) {
        actions.push_back(std::shared_ptr<maf::Action>(new HelloWorldAction(std::string(ii, 's'), 2 * ii, 3.3 * ii, 4.4 * ii * ii)));
    }

    return actions;
}

class TestBcastController : public maf::BcastController {
public:

    void run() override {
        for (auto action : GetActions(3)) {
            this->bcast(action);
        }

        this->bcast(GetActions(3));
    };
};

int main(int argc, char* argv[]) {
    int exit_code = -1;

    try {
        maf::ActionFactory::Register<HelloWorldAction>();

        maf::barrier("===== TestBcastController controller;");

        TestBcastController controller;
        controller.start();

        maf::barrier("===== maf::BcastController controller2(GetActions(3));");

        maf::BcastController controller2(GetActions(3));
        controller2.start();

        maf::barrier("===== maf::ScatterController controller3(GetActions(3));");

        maf::ScatterController controller3(GetActions(3));
        controller3.start();
        exit_code = 0;
    }
    catch (std::exception& ex) {
        maf::log("FAILED: ", ex.what());
        MPI_Abort(MPI_COMM_WORLD, -1);
    }
    catch (...) {
        maf::log("FAILED: no idea what happened");
        MPI_Abort(MPI_COMM_WORLD, -1);
    }

    return exit_code;
}

File: examples/py/example02_hello_world.py
'''test functionality in Python'''
import maf

import sys
import cPickle
import marshal
import timeit

@maf.action
class MafSerializer(maf.Action):

    def __init__(self, my_string='Hello World!', my_int=0, my_float=0.0):
        maf.Action.__init__(self)
        self.my_string = my_string
        self.my_int = my_int
        self.my_float = my_float

    def run(self):
        maf.log('{} my_int: {} my_float: {}'.format(self.my_string, self.my_int, self.my_float))
    
    def serialize(self, archive):
        self.my_string = archive.rw_string(self.my_string)
        self.my_int = archive.rw_long_long(self.my_int)
        self.my_float = archive.rw_double(self.my_float)


@maf.action
class PickleSerialer(MafSerializer):

    def serialize(self, archive):
        data = dict(self.__dict__)
        if 'this' in data:
            del data['this'] # don't pickle the SWIG object
        string = cPickle.dumps(data)
        string = archive.rw_string(string)
        self.__dict__.update(cPickle.loads(string))


@maf.action
class MarshalSerialer(MafSerializer):

    def serialize(self, archive):
        data = dict(self.__dict__)
        if 'this' in data:
            del data['this'] # don't pickle the SWIG object
        string = marshal.dumps(data)
        string = archive.rw_string(string)
        self.__dict__.update(marshal.loads(string))


class TestBcastController(maf.BcastController):
    
    def __init__(self, action_class):
        maf.BcastController.__init__(self)
        self.action_class = action_class

    def run(self):
        act = self.action_class()
        self.bcast(act)
        act2 = self.action_class('Whatever', 2, 44)
        self.bcast(act2)
        act = self.action_class('The final act!', sys.maxint, sys.float_info.max)
        self.bcast(act)

maf.barrier('==== Starting tests with MafSerializer')

start = timeit.default_timer()
controller = TestBcastController(MafSerializer)
controller.start()
serializer_time = timeit.default_timer() - start

maf.barrier('==== Starting tests with PickleSerializer')

start = timeit.default_timer()
controller = TestBcastController(PickleSerialer)
controller.start()
pickle_time = timeit.default_timer() - start

maf.barrier('==== Starting tests with MarshalSerializer')

start = timeit.default_timer()
controller = TestBcastController(MarshalSerialer)
controller.start()
marshal_time = timeit.default_timer() - start

maf.barrier('==== Done with tests')

maf.log(              'SERIALIZE time: {}\n'
    '                  PICKLE time   : {}\n'
    '                  MARSHAL time  : {}'
    .format(serializer_time, pickle_time, marshal_time))
